### PR TITLE
limit the number of retries of the busy handler

### PR DESCRIPTION
**Pull request by MarinPostma**
This PR fixes a series of bugs that caused hangs on the platform.

- Replace `is_stolen` `AtomicBool` by a `Mutex<bool>`. I'm not sure this one was necessary, but I was desperate.
- Always update the state for a connection, even in case of error during execution.
- An instance of `borrow_and_update` in a channel was borrowing too long, causing a dead_lock. This was caused by some obscure rules about temporaries, causing the reference to be dropped after the call, even though it was being dereferenced. This is best exemplified by https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3c7bf7fbc2471c9bce121c781ebc982b, and elements of explanation can be found here: https://blog.m-ou.se/super-let/

### CODE REVIEW

File: libsql-server/src/connection/libsql.rs [hunk 1]
```diff
@@ -621,61 +642,71 @@ impl<W: Wal> Connection<W> {
...
+            // lock was downgraded, notify a waiter
+            (Tx::Write, Tx::None | Tx::Read) => {
+                let old_slot = self
+                    .slot
+                    .take()
+                    .expect("there should be a slot right after downgrading a txn");
+                let mut maybe_state_slot = self.state.slot.write();
+                // We need to make sure that the state slot is our slot before removing it.
+                if let Some(ref state_slot) = *maybe_state_slot {
+                    if Arc::ptr_eq(state_slot, &old_slot) {
```
Comment (LucioFranco) [inline libsql-server/src/connection/libsql.rs:694]: is there a case where a pointer gets reused and we run into a ABA problem?

**Review by LucioFranco — APPROVED** [review]

File: libsql-server/src/connection/libsql.rs (hunk 1 above)
Comment (MarinPostma) [reply libsql-server/src/connection/libsql.rs:694]: no, we discussed that last time, but we're actually comparing the pointer with itself, and we're still holding a reference to both ends while comparing, so no risk of it being reused in the meantime :)

File: libsql-server/src/connection/libsql.rs [hunk 2]
```diff
@@ -493,7 +492,17 @@ unsafe extern "C" fn busy_handler<T: Wal>(state: *mut c_void, retries: c_int) ->
                         if Arc::ptr_eq(s, &slot) {
                             // We check that slot wasn't already stolen, and that their is still a slot.
                             // The ordering is relaxed because the atomic is only set under the slot lock.
```
Comment (haaawk) [inline libsql-server/src/connection/libsql.rs:494]: This comment can probably go away, no?

File: libsql-server/src/connection/libsql.rs [hunk 3]
```diff
@@ -612,12 +621,17 @@ impl<W: Wal> Connection<W> {
...
-                    has_timeout = true;
+            if !has_timeout {
+                if let Some(slot) = &lock.slot {
+                    let mut is_stolen = slot.is_stolen.lock();
+                    if *is_stolen || Instant::now() > slot.expires_at() {
+                        // we mark ourselves as stolen to notify any waiting lock thief.
+                        if !*is_stolen {
+                            lock.rollback();
+                        }
+                        *is_stolen = true;
```
Comment (haaawk) [inline libsql-server/src/connection/libsql.rs:630]: Shouldn't this be in opposite order?
```
let stolen = *is_stolen
*is_stolen = true;
if !stolen {
  lock.rollback();
}
```

**Review by haaawk — APPROVED** [review]

**Review by psarna — APPROVED** [review]
looks good, later we can also push a low priority follow-up that brings back `AtomicBool`, since it's probably just as good -- and test it on canary

**Pull request merged**